crates/aptos-logger/src/aptos_logger.rs

```diff
@@ -1024,4 +1055,135 @@ mod tests {
                 "source_path"
             )));
     }
+
+    // Test speculative logging via SpeculativeEvents, here because the testing infrastructure can be re-used.
+    struct SpeculativeLog {
+        level: Level,
+        message: String,
+    }
+
+    impl SpeculativeLog {
+        fn new(level: Level, message: String) -> Self {
+            Self { level, message }
+        }
+    }
+
+    impl SpeculativeEvent for SpeculativeLog {
+        fn dispatch(&self) {
+            match self.level {
+                Level::Debug => debug!("{}", self.message),
+                Level::Warn => warn!("{}", self.message),
+                Level::Error => error!("{}", self.message),
+                _ => {},
+            }
+        }
+    }
+
+    #[test]
+    fn test_speculative_logging() {
+        let receiver_mutex = set_test_logger(Level::Debug, true);
+        let receiver = receiver_mutex.lock();
+
+        while receiver.try_recv().is_ok() {}
+
+        let speculative_logs = SpeculativeEvents::<SpeculativeLog>::new(2);
+
+        // out of bounds
+        assert_err!(speculative_logs.record(
+            2,
+            SpeculativeLog::new(Level::Trace, "0/trace: A".to_string()),
```

Comment:
Unless we're matching on the string names, I'd probably remove them. They're a bit distracting 😄

Reply:
was matching on string names :-)